workflow/certprof/process.go

```diff
@@ -0,0 +1,200 @@
+package certprof
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/jessepeterson/mdmcommands"
+	"github.com/micromdm/nanocmd/logkeys"
+	"github.com/micromdm/nanocmd/workflow"
+	"github.com/micromdm/nanolib/log/ctxlog"
+)
+
+// Start starts the workflow.
+func (w *Workflow) Start(ctx context.Context, step *workflow.StepStart) error {
+	// load and validate the workflow context
+	wfCtx, ok := step.Context.(*Context)
+	if !ok {
+		return workflow.ErrInvalidContext
+	}
+	if err := wfCtx.Validate(step.Name); err != nil {
+		return fmt.Errorf("validating context: %w", err)
+	}
+
+	// assemble the CertificateList MDM command
+	cmd := mdmcommands.NewCertificateListCommand(w.ider.ID())
+
+	if !wfCtx.NoManagedOnly {
+		managedOnly := true
+		cmd.Command.ManagedOnly = &managedOnly
+	}
+
+	// assemble our StepEnqueuing
+	se := step.NewStepEnqueueing()
+	// add our command
+	se.Commands = []interface{}{cmd}
+	// pass along our context to the subsequent step completions
+	se.Context = wfCtx
+
+	// enqueue the step!
+	return w.enq.EnqueueStep(ctx, w, se)
+}
+
+// replace replaces instances of r's map keys with their values.
+func replace(in string, r map[string]string) string {
+	if len(in) < 1 || r == nil || len(r) < 1 {
+		return in
+	}
+	oldnews := make([]string, 0, len(r)*2)
+	for k, v := range r {
+		oldnews = append(oldnews, k, v)
+	}
+	repl := strings.NewReplacer(oldnews...)
+	return repl.Replace(in)
+}
+
+// filter in using f.
+func filter(f *Filter, in []mdmcommands.CertificateListItem) (out []mdmcommands.CertificateListItem) {
+	if f == nil {
+		return
+	}
+	for _, item := range in {
+		// skip if cert is not an identity
+		if !f.NoRequireIdentity && !item.IsIdentity {
+			continue
+		}
+
+		// skip if no CN prefix match
+		if f.CNPrefix != "" && !strings.HasPrefix(item.CommonName, f.CNPrefix) {
+			continue
+		}
+
+		out = append(out, item)
+	}
+	return
+}
+
+// checkCriteria compares item against c to determine whether it should be replaced (re-installed).
+func checkCriteria(c *Criteria, item *mdmcommands.CertificateListItem) (replace bool, reason string, err error) {
+	if c == nil {
+		// if there is no criteria then the mere existence of the cert
+		// means we should never replace it.
+		return false, "nil criteria", nil
+	}
+
+	if c.AlwaysReplace {
+		return true, "always replace", nil
+	}
+
+	if c.UntilExpirySeconds > 0 {
+		crt, err := x509.ParseCertificate(item.Data)
+		if err != nil {
+			return true, "error", fmt.Errorf("parse certificate: %w", err)
+		}
+
+		if time.Until(crt.NotAfter) < (time.Duration(c.UntilExpirySeconds) * time.Second) {
+			return true, fmt.Sprintf("certificate NotAfter Until: %d; less than expiry: %d", time.Until(crt.NotAfter), c.UntilExpirySeconds), nil
```

Comment:
Maybe add "replace" so its clear that it is replacing it, similar to the other cases.

Reply:
I got to here for this review. I'll do another when I have a chunk of time tonight.